DOC clarifications on the release process

adrinjalali opened the pull request:
adding a few clarifications to the release process.

adrinjalali committed:
some extra comments

adrinjalali committed:
reorganize steps

adrinjalali committed:
more info

adrinjalali committed:
more clarification

adrinjalali committed:
further instructions

adrinjalali committed:
be cautious with tagging

adrinjalali committed:
add pypi note

adrinjalali committed:
git log copy

adrinjalali committed:
upstream

adrinjalali committed:
more explanation

adrinjalali committed:
fix typo

adrinjalali committed:
clarify

adrinjalali committed:
typo

adrinjalali commented:
@jnothman anything I have as mistakes here? @NicolasHug WDYT? I'm happy to clarify wherever's not clear.

NicolasHug commented on doc/developers/maintainer.rst:
```diff
@@ -1,14 +1,24 @@
 
+A release is either a major release, or a minor bug fix release. Our convention
+is that we release one or more release candidates (0.XXrcN) before releasing
+the final binaries. Creating the first RC involves creating the release branch
```
```suggestion
the final distributions. Creating the first RC involves creating the release branch
```

since we also ship source dists?

NicolasHug commented on doc/developers/maintainer.rst:
```diff
@@ -17,25 +27,30 @@ Before a release
+Since any commits to an existing branch (e.g. 0.999.X) will automatically
+update the web site documentation, it is best to develop a release with a pull
 request in which 0.999.X is the base. It also allows you to keep track of any
 tasks towards release with a TO DO list.
```
I don't understand this, what documentation is updated exactly, and why is it a bad thing?

adrinjalali replied:
does this help?

NicolasHug commented on doc/developers/maintainer.rst:
```diff
@@ -17,25 +27,30 @@ Before a release
+    $ git rebase -i upstream/0.999.X
+
+It's nice to have a copy of the ``git rebase -i`` log in the PR to help others
+understand what's included.
```
Also explain what should and what should not be included during the rebase? My understanding is that:

- to create e.g. 0.21.rc0, we include everything from 0.20.X 
- to create e.g. 0.21.rc1, we only select specific commits from 0.21.rc0. I.e. we include additional bug fixes and minor docs but typically ignore new features that weren't already in rc0.

NicolasHug reviewed (commented):
A few comments, thanks for all the work @adrinjalali

adrinjalali committed:
further clarifications, address comments

cmarmo commented on doc/developers/maintainer.rst:
```diff
@@ -1,14 +1,24 @@
+branch (0.99.X for instance) directly on the main repo. The rest of the changes
+should be done as a PR to the master branch, if it's possible, or done through
+a PR to the release branch, for the other maintainers to review before the
+release. This is further explained under *Preparing a release PR*.
```
@adrinjalali one may expect a link to the corresponding subsection here? 
```suggestion
release. This is further explained under :ref:`Preparing a release PR<Preparing a release PR>`.
```

NicolasHug commented on doc/developers/maintainer.rst:
```diff
@@ -17,25 +27,54 @@ Before a release
+
+Since any commits to an existing branch (e.g. 0.999.X) will automatically
+update the web site documentation, it is best to develop a release with a pull
 request in which 0.999.X is the base. It also allows you to keep track of any
```
Not sure I understand what "0.999.X is the base" means.  does it mean that the PR tries to merge things into the 0.999.X branch? Is there any other way to do it anyway?

adrinjalali replied:
For me the two git commands a few lines later explain what it is. I also didn't understand what it meant from the text.

cmarmo commented:
As external observer I would like to read here something explaining the different types of release (like [the python one](https://www.python.org/dev/peps/pep-0101/#types-of-releases) for example).... obviously this can wait another PR.

NicolasHug commented on doc/developers/maintainer.rst:
```diff
@@ -17,25 +27,54 @@ Before a release
+......................
+
+Since any commits to an existing branch (e.g. 0.999.X) will automatically
+update the web site documentation, it is best to develop a release with a pull
```
Is it true that any branch to the repo has an online documentation? I thought we'd have a way to curate which branches we want to have the docs for, like with ReadTheDocs

adrinjalali replied:
AFAIK any `0.XX.X` branch has a documentation online. Any change to the `0.21.X` branch for example, will be there almost immediately. And to add a new release, you just create the branch, and it'll automatically be there.

adrinjalali commented:
> As external observer I would like to read here something explaining the different types of release (like the python one for example).... obviously this can wait another PR.

That's a good point. I also read that one while setting the tags. I left a link to it.

adrinjalali committed:
add link to pep101

NicolasHug reviewed (commented):
Trying to summarize the paragraph that I don't understand:

```
First, create a branch (to release e.g. 0.999.3):

    $ # assuming master and upstream/master are the same
    $ git checkout -b release-0.999.3 master

Note that pushing the (empty) branch will 
automatically upload its documentation, which currently 
corresponds to that of master, so it's not an accurate representation
of what the release will actually be.

Then create a PR with all the desired changes:

	$ git rebase -i upstream/0.999.2

Ideally, the PR should be merged quickly, to minimize the 
gap between the time when the website points to 
a release, and the release
being available to users.
```

jnothman commented on doc/developers/maintainer.rst:
```diff
@@ -54,24 +97,19 @@ Making a release
+   when ready to release. On master, increment the version in the same place
+   (when branching for release). This means while we're in the release
+   candidate period, the latest stable is two versions behind the master
+   branch, and one one.
```
One one was a racehorse two two was one two. One one one one race. Two two one one two. 

Not sure what one one is doing here.

jnothman commented on doc/developers/maintainer.rst:
```diff
@@ -81,10 +119,31 @@ Making a release
    that the tests pass.
 
+4. Proceed with caution. Ideally, tags should be created when you're almost
+   certain that the release is ready. You can test upload the sdist to
```
To avoid premature tagging,...

jnothman commented on doc/developers/maintainer.rst:
```diff
@@ -81,10 +119,31 @@ Making a release
 
+4. Proceed with caution. Ideally, tags should be created when you're almost
+   certain that the release is ready. You can test upload the sdist to
+   test.pypi.org, and follow step 5 and have ``BUILD_COMMIT`` set to the
```
```suggestion
   test.pypi.org, and test the next step by setting ``BUILD_COMMIT`` to the
```

jnothman commented on doc/developers/maintainer.rst:
```diff
@@ -81,10 +119,31 @@ Making a release
+4. Proceed with caution. Ideally, tags should be created when you're almost
+   certain that the release is ready. You can test upload the sdist to
+   test.pypi.org, and follow step 5 and have ``BUILD_COMMIT`` set to the
+   branch name (``0.22.X`` for instance), and test the wheel builds on a PR on
```
```suggestion
   branch name (``0.22.X`` for instance) in a PR to the wheel building repo.

jnothman commented on doc/developers/maintainer.rst:
```diff
@@ -119,8 +182,8 @@ Making a release
-       $ sed -i "s/latestStable = '.*/latestStable = '0.999';" versionwarning.js
-       $ git commit -m "Update stable to point to 0.999" stable
+       $ sed -i "s/latestStable = '.*/latestStable = '0.999';/g" versionwarning.js
+       $ git commit -am "Update stable to point to 0.999"
```
I don't get why you prefer this to explicitly only committing stable

adrinjalali replied:
the `stable` used to be the only thing changing. Now `versionwarning.js` also changes in this step.

rth replied:
Added `git add stable/ versionwarning` on the line above to be explicit.

jnothman commented on doc/developers/maintainer.rst:
```diff
@@ -119,8 +182,8 @@ Making a release
        $ ln -s 0.999 stable
-       $ sed -i "s/latestStable = '.*/latestStable = '0.999';" versionwarning.js
-       $ git commit -m "Update stable to point to 0.999" stable
+       $ sed -i "s/latestStable = '.*/latestStable = '0.999';/g" versionwarning.js
```
is g necessary?

adrinjalali replied:
yeah, my `sed` complained about an unfinished `s/` directive or something w/o it.

jnothman replied:
You're right. It needs the / but not the g

jnothman reviewed (commented):
Thank you @adrinjalali!

cmarmo commented on doc/developers/maintainer.rst:
```diff
@@ -1,14 +1,27 @@
+should be done as a PR to the master branch, if it's possible, or done through
+a PR to the release branch, for the other maintainers to review before the
+release. This is further explained under
+:ref:`Preparing a release PR<Preparing a release PR>`. We follow the
```
... me asking for forgiveness for the sphinx warning... this time I've checked
 ```suggestion
`Preparing a release PR<#Preparing-a-release-PR>`_. We follow the
```

adrinjalali replied:
haha, thanks for the fix.

adrinjalali committed:
explain branch names

adrinjalali commented on doc/developers/maintainer.rst:
```diff
@@ -17,25 +30,54 @@ Before a release
+is live as a stable release, it should ideally match the distributions
+available to users. Having all the changes for a release in one PR minimizes
+the gap between the time when the website points to a release, and the release
+being available to users.
```
yes, having one PR just makes it clearer what the difference between two releases are.

NicolasHug replied:
So the current description should be updated. Right now it says have only one PR helps minimizing the time gap, which isn't true

adrinjalali committed:
apply comments

adrinjalali commented:
In the first paragraph, I added how the branches are named, does that help clarify the confusing paragraph @NicolasHug ?

NicolasHug commented on doc/developers/maintainer.rst:
```diff
@@ -17,25 +30,54 @@ Before a release
+    $ # # assuming master and upstream/master are the same
     $ git checkout -b release-0.999.3 master
-    $ git rebase -i 0.999.X
+    $ git rebase -i upstream/0.999.X
```
what branch name? the branch we're creating here is named release-0.999.3

NicolasHug commented on doc/developers/maintainer.rst:
```diff
@@ -1,14 +1,31 @@
 ========================================
 
+A major release, incrementing the minor version, or a big fix release
+incrementing the patch version. Our convention is that we release one or more
```
that sentence is missing something

also big -> bug?

NicolasHug commented on doc/developers/maintainer.rst:
```diff
@@ -1,14 +1,31 @@
+A major release, incrementing the minor version, or a big fix release
+incrementing the patch version. Our convention is that we release one or more
+release candidates (0.XXrcN) before releasing the final distributions. Creating
+the first RC involves creating the release branch (0.99.X for instance)
```
using X for 2 different things here?

NicolasHug commented on doc/developers/maintainer.rst:
```diff
@@ -1,14 +1,31 @@
+release candidates (0.XXrcN) before releasing the final distributions. Creating
+the first RC involves creating the release branch (0.99.X for instance)
+directly on the main repo. The rest of the changes should be done as a PR to
+the master branch, if it's possible, or done through a PR to the release
```
what are "the rest of the changes?" in this context?

NicolasHug reviewed (commented):
Made a few comments, but sorry  I still find the paragraph confusing.

Maybe comment on what isn't correct about my suggestion in https://github.com/scikit-learn/scikit-learn/pull/15759#pullrequestreview-326289473 (this is my current understanding of things)

cmarmo commented on doc/developers/maintainer.rst:
```diff
@@ -1,14 +1,31 @@
+happens under the branch called ``0.99.X``, where ``X`` is a literal, and
+independent of minor bug fix releases. Each release candidate, major, or minor
+release is a tag under the same branch. This is further explained under
+`Preparing a release PR<#Preparing-a-release-PR>`_. We follow the `PEP101
```
shame on me....
line 15 should be
```
:ref: `preparing-a-release-pr`. We follow the `PEP101`
```
Then a label should be added before the section:
```
.. _preparing-a-release-pr:
```
:(

adrinjalali replied:
no worries @cmarmo, I'm actually quite happy you figured it out. I'm not the most fluent when it comes to ReST as you can see :P

glemaitre commented on doc/developers/maintainer.rst:
```diff
@@ -1,14 +1,31 @@
+incrementing the patch version. Our convention is that we release one or more
+release candidates (0.XXrcN) before releasing the final distributions. Creating
+the first RC involves creating the release branch (0.99.X for instance)
+directly on the main repo. The rest of the changes should be done as a PR to
```
Could we call this stage "Freeze". Supposidely, we should not add API change, new feature after this stage.

adrinjalali committed:
cover more comments

adrinjalali commented:
@NicolasHug I think adding this modifications to your text makes it correct (all in **bold**):


First, create a branch, **on your own fork** (to release e.g. 0.999.3):

    $ # assuming master and upstream/master are the same
    $ git checkout -b release-0.999.3 master

Note that pushing the (empty) branch **to the main repo** will 
automatically upload its documentation, which currently 
corresponds to that of master, so it's not an accurate representation
of what the release will actually be.

Then create a PR **to the `scikit-learn/0.999.X` branch and not `master`**
with all the desired changes:

	$ git rebase -i upstream/0.999.2

Ideally, **once the PR is merged, the new tag should be soon after
created and wheels uploaded**, to minimize the gap between the
time when the website points to a release, and the release
being available to users.
```

NicolasHug commented:
I would simply rephrase the last sentence with

> The rest of the process is to merge the PR, create a tag and publish the wheels. Ideally, this should be done quickly in order to minimize the gap between...

but otherwise I'm happy with it :)

question: is the branch merged into master at some point? If not, how do we sync both branches when some changes have been directly made to the branch but not to master?

qinhanmin2014 commented:
I think we need to mention #15840 before we figure out a better way.

jnothman commented:
The only changes that should be directly made to the release branch are
version number changes. Otherwise all commits should be picked across from
changes in master. Occasionally it lands up going the other direction.

adrinjalali committed:
clarify branching, remove Dash ref

adrinjalali commented:
I clarified the branching, added the binder reference, and removed the dash doc reference. I don't think they have a healthy community, and I don't see why we should contribute to it. Obviously we're all free to contribute to that repo if we want to, but it certainly shouldn't be an expectation from the release manager.

jnothman commented:
Re dash: It wasn't an expectation on a release manager, only a reminder to
myself. They don't have a developer community as such. They have a
developer, and a user community. I think the developer could set
expectations more clearly and has rejected my attempts to help do so. And I
think that his style of interaction on pull requests is different to ours
and was brutally unfamiliar to you. Yet if we automate more of this
deployment pipeline, we should consider automating that too.

adrinjalali commented:
So otherwise are we okay with the sate of this PR now? we can always improve later.

qinhanmin2014 committed:
Update maintainer.rst

qinhanmin2014 commented:
I've updated the binder PR.

adrinjalali committed:
merge upstream/master

rth commented on doc/developers/maintainer.rst:
```diff
@@ -17,25 +36,70 @@ Before a release
+understand what's included.
+
+In terms of including changes, the first RC is a pure copy of the ``master``
+with a release version number, and ideally counts as a *feature freeze*. Each
```
I don't think it's a helpful technical detail. One can create the release branch, try to build wheels, see something that fails, add commits on top to fix it etc.  Saying that the first RC should be the exact copy of master is not true in general.

rth commented on doc/developers/maintainer.rst:
```diff
@@ -81,10 +141,31 @@ Making a release
+
+   - You can test if PyPi is going to accept the package using::
+
+       $ twine check dist/*
```
Not sure about the point of building the wheel locally and running `twine check` on it. I think using only wheels built in CI is less prone to confusion.

adrinjalali replied:
We still have to build the `sdist` locally anyway. So we should check that one at least locally. Also, the build CI is pretty slow, and doing a check before triggering that is not a bad idea.

rth commented on doc/developers/maintainer.rst:
```diff
@@ -81,10 +141,31 @@ Making a release
+
+       $ twine check dist/*
+
+   You can run ``twine check`` after step 5 (fetching artifacts) as well.
```
Let's move it there instead then.

adrinjalali replied:
I put it here cause step 5 is after creating the tag. Ideally once the tag is created, there shouldn't be more necessary changes.

rth commented on doc/developers/maintainer.rst:
```diff
@@ -81,10 +141,31 @@ Making a release
    that the tests pass.
 
+4. Proceed with caution. Ideally, tags should be created when you're almost
+   certain that the release is ready to avoid premature tagging. You can test
```
```suggestion
   certain that the release is ready. You can test
```
as it's redundant otherwise..

rth commented on doc/developers/maintainer.rst:
```diff
@@ -81,10 +141,31 @@ Making a release
+
+    $ git tag -a 0.999
+
+    $ git push git@github.com:scikit-learn/scikit-learn.git --tags
```
No -- you can have multiple unrelated tags locally.
```suggestion
    $ git push git@github.com:scikit-learn/scikit-learn.git 0.999
```

adrinjalali replied:
I actually do steps 3-6 on a different clone where `origin` is the main repo, and is clean of all branches tags of my own fork, should I put that in?

rth replied:
I pushed a commit to explicitly indicate the pushed tag. I would not be comfortable using the above command as there is always a risk of making an error on the repo in one is etc.

rth commented on doc/developers/maintainer.rst:
```diff
@@ -17,25 +36,70 @@ Before a release
+to avoid asynchrony. First, create a branch, **on your own fork** (to release
+e.g. `0.999.3`)::
 
+    $ # assuming master and upstream/master are the same
```
For me `origin/master` is my fork, `upstream/master` is official scikit-learn master. I guess that's not what this means? I know different people have different conventions. Maybe better to avoid talking about `upstream` to avoid confusion if so. Not that in github docs `upstream` is also not one's fork https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork

adrinjalali replied:
yes, I exactly meant `upstream` with the same semantics as you mention.

rth commented on doc/developers/maintainer.rst:
```diff
@@ -17,25 +36,70 @@ Before a release
+Then create a PR **to the `scikit-learn/0.999.X` branch and not `master`** with
+all the desired changes::
+
+	$ git rebase -i upstream/0.999.2
```
So this is this a branch in the fork or in the official scikit-learn repo? The documentation doesn't say how it was created..

adrinjalali replied:
I added a step 0

rth reviewed (commented):
A few minor comments, LGTM otherwise. Thanks @adrinjalali !

adrinjalali committed:
address some of Roman's comments

NicolasHug commented on doc/developers/maintainer.rst:
```diff
@@ -1,14 +1,33 @@
+maintainers to review before the release. Note that all development for the
+minor releases and the major release of for instance 0.99, happens under the
+branch called ``0.99.X``, where ``X`` is a literal, and independent of minor
+bug fix releases. Each release candidate, major, or minor release is a tag
```
I don't understand this sentence, especially since it is said above that changes should be done to master whenever possible

adrinjalali replied:
The development happens on that branch, through PRs which are merged into that branch.

NicolasHug commented on doc/developers/maintainer.rst:
```diff
@@ -17,25 +36,69 @@ Before a release
    - ``maint_tools/sort_whats_new.py`` can put what's new entries into
-     sections.
+     sections. It's not perfect, and requires manual checking of the changes.
+     If the whats new list if well curated, it may not be necessary.
```
```suggestion
     If the whats new list is well curated, it may not be necessary.
```

NicolasHug commented on doc/developers/maintainer.rst:
```diff
@@ -17,25 +36,69 @@ Before a release
-happen in master to avoid asynchrony. To select commits from master for use in
-the bug fix (version 0.999.3), you can use::
+Most development of the release, and its documentation, should happen in master
+to avoid asynchrony. First, create a branch, **on your own fork** (to release
```
This whole  first paragraph and the first sentence of the second one feel redundant with the rest

adrinjalali replied:
Not sure how to fix that.

NicolasHug replied:
I have suggested an alternative paragraph before

adrinjalali committed:
address some of Nocolas's suggestions

NicolasHug commented on doc/developers/maintainer.rst:
```diff
@@ -1,14 +1,33 @@
+branch, for the other maintainers to review before the release. Note that all
+development for the minor releases and the major release of for instance 0.99,
+happens under the branch called ``0.99.X``, where ``X`` is a literal, and
+independent of minor bug fix releases. Each release candidate, major, or minor
```
I still think this is ambiguous given the previous sentence. Please clarify what "development happens" means or change the phrasing?

adrinjalali committed:
trying to address comments

adrinjalali commented:
I tried to address them. If they're still not clear, feel free to apply changes directly. Don't really feel like working on this PR for much longer :D

thomasjpfan committed:
CLN Adds reference to conda-forge recipe/meta.yml

cmarmo reviewed (approved):
@adrinjalali even if I'm not a core dev may I approve? I feel as never ending reviews are contraindicated for core-devs too, not only for beginners... :)
The best review ever for this PR will be the next release. WDYT?

rth reviewed (approved):
Thanks @cmarmo ! Yes let's merge this, it been a quite long and detailed discussed for a documentation change that only impacts the release manager (who will likely be able to detect if anything is amiss while going through the instructions).

I think most comments are addressed, waiting 1 day to see if there are any objections.

NicolasHug commented:
let's not merge yet. Last time I looked it wasn't clear to me, and I haven't had to time to check again.

There's no rush anyway

rth committed:
DOC Explicitly push the tag by name

rth committed:
DOC Explicitly git add folders before commiting

rth commented:
OK I added a few minor fixed, IMO it's good to go.

> let's not merge yet. Last time I looked it wasn't clear to me, and I haven't had to time to check again.

It's been 1.5 months, 5 reviewers and 81 comments, for something that was expected to be just Adrin documenting his experience as release manager. Never ending reviews are frustrating both for PR authors and reviewers. You were mostly fine with it in December i think https://github.com/scikit-learn/scikit-learn/pull/15759#issuecomment-563251839 ?

I would rather we merged it and made follow up clarifications if needed. In any case I see this as a rough guideline for the release manager, and would expect them to have an overall picture of what needs to be done, be sure they understand each of those commands before executing them and asking questions/making clarification PRs if there is any doubt.

NicolasHug commented:
> You were mostly fine with it in December i think #15759 (comment) ?

I was only referring to Adrin's modifications to my suggestion further up, not to the PR as a whole.

I don't think there's any urgency here, this is not even a user-facing doc. I'll try to review and push my own stuff soon (by the end of next week)

NicolasHug committed:
fixed ref and removed redundant sentence (I think)

NicolasHug committed:
Added note about making sure derecations are handled

NicolasHug committed:
fixed title headers, this file is not just about releasing

NicolasHug committed:
removed redundant paragraph

NicolasHug committed:
major changes

NicolasHug commented on doc/developers/maintainer.rst:
```diff
@@ -81,10 +128,32 @@ Making a release
+   ``test.pypi.org``, and test the next step by setting ``BUILD_COMMIT`` to the
+   branch name (``0.22.X`` for instance) in a PR to the wheel building repo.
+   Once all works, you can proceed with tagging. Create the tag and push it (if
+   it's an RC, it can be ``0.xxrc1`` for instance)::
```
I understand the tagging should be done in the 0.99.X branch?

adrinjalali replied:
yes, correct.

NicolasHug commented on doc/developers/maintainer.rst:
```diff
@@ -17,61 +30,95 @@ Before a release
 
 Making a release
-----------------
+................
```
At what point do we merge the PR mentioned above into the 0.99.X branch? Should we add a step in the list below?

adrinjalali replied:
It has to happen before tagging. You can add one if you think it's not clear.

NicolasHug reviewed (commented):
I have pushed some changes. I still have a few questions left, please take a look when you can @adrinjalali

adrinjalali commented:
> I don't think there's any urgency here, this is not even a user-facing doc. I'll try to review and push my own stuff soon (by the end of next week)

You could also merge and have your changes in a separate PR :P which would make it easier for me and others to review. I'm kinda lost now :D 

Have you removed the part where I talked about why the release should ideally happen very quickly after the PR is merged? I think that's an important point if it's missing now.

NicolasHug commented:
> Have you removed the part where I talked about why the release should ideally happen very quickly after the PR is merged? I think that's an important point if it's missing now.

Yes I removed it because I could not make sense out of it. The cause to effect relationships were broken somewhere. Here is what it was

https://github.com/scikit-learn/scikit-learn/blob/533a122521dfa85909ec843515f9818a7e66ff30/doc/developers/maintainer.rst#preparing-a-release-pr

In particular, the sentence "Note that pushing the (empty) branch **to the main repo** will automatically upload its documentation" is confusing because right above the branch is created on the fork. So it's not clear what branch this is referring to.

> You could also merge and have your changes in a separate PR :P which would make it easier for me and others to review

Feel free to merge, I pushed because you suggested to before

NicolasHug committed:
minor comments

NicolasHug committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into pr-15759

adrinjalali commented:
Thank you all for the commits and comments. Merging this one, happy to have further improvements on it.
